File: functions/src/features/quests/submit-answer.ts
import {HttpsError, onCall} from "firebase-functions/v2/https";
import {getSignedInUser} from "../../shared/get_signed_in_user";
import {Quest} from "./types/quest";
import {getFirestore, Timestamp} from "firebase-admin/firestore";
import {Points} from "../points/types/points";
import {PointsTypeEnum} from "../points/types/points-type-enum";
import {logger} from "firebase-functions/v2";

export const submitAnswer = onCall(
  {region: "europe-west3"},
  async (request) => {
    const loggedUser = await getSignedInUser(request);
    const questId = request.data.quest as string;
    const answers = request.data.answers as number[];
    const questSnap = await getFirestore()
      .collection("quests")
      .doc(questId)
      .get();
    if (!questSnap.data()) {
      throw new HttpsError("not-found", "Quest not found");
    }
    const quest = <Quest>{...questSnap.data(), id: questSnap.id};

    logger.info("Quest answers: " + quest.id);

    const correctAnswers = quest.answers
      ?.filter((answer) => answer.isCorrect)
      .map((answer) => answer.id);

    if (!correctAnswers) {
      throw new HttpsError("not-found", "Quest has no correct answers");
    }

    logger.info("Correct answers: " + correctAnswers);

    const isCorrect = correctAnswers.every((e) => answers.includes(e));

    logger.info("Is correct: " + isCorrect);

    const archivedPoints = <Points>{
      type: PointsTypeEnum.quest,
      points: quest.points,
      assignedAt: Timestamp.now(),
    };

    const batch = getFirestore().batch();
    if (isCorrect) {
      batch.set(
        getFirestore()
          .collection("users")
          .doc(loggedUser.uid)
          .collection("points")
          .doc(quest.id),
        archivedPoints,
      );
    }

    batch.update(
      getFirestore().collection("users").doc(loggedUser.uid),
      {
        activeQuest: null,
      },
    );

    await batch.commit();

    return isCorrect;
  });

File: functions/features/points/submit_answer.py
from firebase_functions.https_fn import on_call, CallableRequest, HttpsError, FunctionsErrorCode
from google.cloud.firestore import SERVER_TIMESTAMP

from features.points.types.points import Points
from features.points.types.points_type_enum import PointsTypeEnum
from features.quests.types.quest import Quest
from features.quests.types.answer import Answer
from shared.get_signed_in_user import get_signed_in_user
from firestore_client import client as firestore_client
from logger_config import logger
from shared.env import FIREBASE_REGION


@on_call(region=FIREBASE_REGION)
def submit_answer(request: CallableRequest) -> bool:
    logged_user = get_signed_in_user(request)

    quest_id = request.data.get("quest")
    answers = [int(ans) for ans in request.data.get("answers", [])]

    quest_snap = (firestore_client
                  .collection("quests")
                  .document(quest_id)
                  .get())
    if not quest_snap.exists:
        raise HttpsError(code=FunctionsErrorCode.NOT_FOUND, message="Quest not found")

    quest_data = quest_snap.to_dict()
    quest_data["id"] = quest_snap.id

    if quest_data.get("answers", None) is not None:
        quest_data["answers"] = [Answer(**ans) for ans in quest_data["answers"]]

    quest = Quest.model_validate(quest_data)

    logger.info(f"Quest answers: {quest.id}")

    correct_answers = []
    if quest.answers is not None:
        correct_answers = [answer["id"] for answer in quest.answers if answer.get("isCorrect")]

    if len(correct_answers) == 0:
        raise HttpsError(code=FunctionsErrorCode.NOT_FOUND, message="Quest has no correct answers")

    logger.info(f"Correct answers: {correct_answers}")

    is_correct = all(e in answers for e in correct_answers)
    logger.info(f"Is correct: {is_correct}")

    archived_points = Points(
        type=PointsTypeEnum.quest, 
        quest=quest.id,
        assignedBy=logged_user.email,
        points=quest.points[0],
        assignedAt=SERVER_TIMESTAMP
    )

    batch = firestore_client.batch()
    if is_correct:
        doc_ref = (
            firestore_client
            .collection("users")
            .document(logged_user.uid)
            .collection("points")
            .document(quest.id)
        )

        batch.set(doc_ref, archived_points.model_dump())

    user_ref = firestore_client.collection("users").document(logged_user.uid)
    batch.update(user_ref, {"activeQuest": None})

    batch.commit()

    return is_correct
